spatialpandas/geometry/base.py

```diff
@@ -557,14 +626,60 @@ def _from_sequence(cls, scalars, dtype=None, copy=None):
     def take(self, indices, allow_fill=False, fill_value=None):
         from pandas.core.algorithms import take
 
-        data = self.astype(object)
-        if allow_fill and fill_value is None:
-            fill_value = self.dtype.na_value
-        # fill value should always be translated from the scalar
-        # type for the array, to the physical storage type for
-        # the data, before passing to take.
-        result = take(data, indices, fill_value=fill_value, allow_fill=allow_fill)
-        return self._from_sequence(result, dtype=self.dtype.subtype)
+        if allow_fill:
+            # fill value should always be translated from the scalar
+            # type for the array, to the physical storage type for
+            # the data, before passing to take.
+            if fill_value is None:
+                fill_value = self.dtype.na_value
+
+            data = self.astype(object)
+            result = take(data, indices, fill_value=fill_value, allow_fill=allow_fill)
+            return self._from_sequence(result, dtype=self.dtype.subtype)
```

Comment:
I think you should be able to use pyarrow's take as well here. Pandas uses -1 in the indices to indicate missing values in the take, while pyarrow uses nulls in the indices. So if you do that conversion, I think the pyarrow take should do the required thing.